Sources/Datadog/Core/Feature.swift

```diff
@@ -32,11 +32,17 @@ internal struct FeaturesCommonDependencies {
 }
 
 internal struct FeatureStorage {
-    /// Writes data to files.
+    /// Writes data to files. This `Writer` takes current value of the `TrackingConsent` into consideration
+    /// to decided if the data should be written to authorized or unauthorized folder.
     let writer: Writer
-    /// Reads data from files.
+    /// Reads data from files in authorized folder.
     let reader: Reader
 
+    /// An arbitrary `Writer` which always writes data to authorized folder.
+    /// Should be only used by components which implement their own consideration of the `TrackingConsent` value
```

Comment:
There is an alternative solution that I have considered, but it was "too much for this" IMO.

Each public feature (Logging / Tracing / RUM) has it's own `Writer` interface. This interface is implemented by `ConsentAwareDataWriter` which considers the `TrackingConsent` before sending data further to one of two internal `Writers`: `authorisedWriter` / `unauthorisedWriter`.

Current approach is based on exposing the `authorisedWriter` to `CrashReporter`.

An alternative that I consider was to implement `CrashReportingFeatureStorage` that exposes `loggingWriter: Writer?` and `rumWriter: Writer?`. Depending on which writer is present, the `CrashReporter` may write data to one of them. The trick would be to synchronise writes done through Crash Reporting `loggingWriter` with writes performed to Logging `writer`  (same for RUM). That could be done by sharing the `FilesOrchestrator` between Logging and Crash Reporting (same for RUM and Crash Reporting).

From that two alternatives, I considered the first one as simpler. We can however keep the second one in mind, to shape the codebase further if we find this practical. WDYT?

Reply:
Yeah that's why I mentioned that it would be out of scope. Alternatives are going to be non trivial and maybe not worth it for now.

So in the first proposition, you compose a couple of Writers together ahead of the DataProcessor, instead of having the Integration grab the right kind of Writer.
In the second proposition, you encapsulate the access to / use of `unauthorisedWriter` inside the `rumWriter`? I don't see where the consent would be checked ahead of calling `unauthorisedWriter` though.

Anyway, both of your proposal are improvements of the Data flow, I like that! Those are higher-level solutions compared to a vague idea I had.
My idea was just to put an extra step to access the Storage's `arbitraryAuthorizedWriter`. Maybe with the Visitor pattern we would statically signal the intent of a class to consider the `TrackingConsent` and access the `arbitraryAuthorizedWriter`.
It is maybe too much but something like:

```
// We don't need to use the term "visit", it could be more specific, it's just to show the idea
protocol ArbitraryWriterVisitor {
    func visit(_ writer: Writer)
}

class CrashReporter: ArbitraryWriterVisitor {
    func visit(_ writer: Writer)
        // check tracking consent
        // use writer to write directly
    }
}

struct FeatureStorage {
    private let arbitraryAuthorizedWriter: Writer

    func accept(_ visitor: ArbitraryWriterVisitor) {
        visitor.visit(arbitraryAuthorizedWriter)
    }
}
```

Again that's out of scope I think, but otherwise your solutions are higher level and I would prefer that kind of approach.

Reply:
The visitor pattern could also be a solution 👌. It will help keeping details in the Crash Reporting module, instead unrevealing it on all features. This will however require revealing more stuff on `DataProcessorFactory` so that the visitor can produce the writer for `.granted` consent.

I think we can keep all this written here and re-visit it if more complexity would have to come into this part of the codebase.

Reply:
I agree. This doesn't need to be addressed right now and an eventual re-visit could look different it would be fine.